datastore/store.go

```diff
@@ -11,29 +12,114 @@ import (
 	"github.com/spacemeshos/go-spacemesh/sql/atxs"
 	"github.com/spacemeshos/go-spacemesh/sql/ballots"
 	"github.com/spacemeshos/go-spacemesh/sql/blocks"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
 	"github.com/spacemeshos/go-spacemesh/sql/poets"
 	"github.com/spacemeshos/go-spacemesh/sql/proposals"
 	"github.com/spacemeshos/go-spacemesh/sql/transactions"
 )
 
 const (
-	atxHdrCacheSize = 600
+	atxHdrCacheSize      = 600
+	malfeasanceCacheSize = 1000
 )
 
 // CachedDB is simply a database injected with cache.
 type CachedDB struct {
 	*sql.Database
 	logger      log.Log
-	atxHdrCache AtxCache
+	atxHdrCache *Cache[types.ATXID, types.ActivationTxHeader]
+
+	// used to coordinate db update and cache
+	mu               sync.Mutex
+	malfeasanceCache *Cache[types.NodeID, types.MalfeasanceProof]
 }
 
 // NewCachedDB create an instance of a CachedDB.
 func NewCachedDB(db *sql.Database, lg log.Log) *CachedDB {
 	return &CachedDB{
-		Database:    db,
-		logger:      lg,
-		atxHdrCache: NewAtxCache(atxHdrCacheSize),
+		Database:         db,
+		logger:           lg,
+		atxHdrCache:      NewAtxCache(atxHdrCacheSize),
+		malfeasanceCache: NewMalfeasanceCache(malfeasanceCacheSize),
+	}
+}
+
+func (db *CachedDB) MalfeasanceCacheSize() int {
+	return db.malfeasanceCache.lru.Len()
+}
+
+// IsMalicious returns true if the NodeID is malicious.
+func (db *CachedDB) IsMalicious(id types.NodeID) (bool, error) {
+	if id == types.EmptyNodeID {
+		return false, errors.New("cache db: get empty node id")
+	}
```

Comment:
Why special-case it?

Reply:
this is a programming error. maybe i should log.fatal this...